mlir/lib/Dialect/MIOpen/Transforms/ApplyImpl.cpp

```diff
@@ -37,53 +38,63 @@ struct MIOpenApplyImplPass
     ModuleOp mod = getOperation();
-    llvm::StringRef modName = miopen::MIOpenDialect::kKernelModuleName;
-    if (mod.getName() == modName)
-      return;
-
-    if (auto miopenMod = mod.lookupSymbol<ModuleOp>(modName)) {
-      SymbolTable symbolTable(mod);
-      auto *ctx = mod.getContext();
-      OpBuilder b(ctx);
-
-      SmallVector<gpu::GPUModuleOp, 8> gpuMods;
-      miopenMod->walk([&](gpu::GPUModuleOp gpuMod) {
-        auto binaryAttr = gpuMod->getAttrOfType<StringAttr>(
-            gpu::getDefaultGpuBinaryAnnotation());
-        if (!binaryAttr) {
-          gpuMod.emitOpError() << "missing gpu.binary attribute";
-          return;
-        }
+    OpBuilder b(mod.getContext());
+
+    llvm::SmallVector<ModuleOp, 8> kernelMods;
+    llvm::SmallDenseMap<func::FuncOp, llvm::SmallVector<DictionaryAttr, 4>>
+        kernelImpls;
+
+    mod->walk([&](ModuleOp kernelMod) {
+      if (kernelMod->hasAttr("kernel.module")) {
+
+        SmallVector<gpu::GPUModuleOp, 8> gpuMods;
+        kernelMod->walk([&](gpu::GPUModuleOp gpuMod) {
+          auto binaryAttr = gpuMod->getAttrOfType<StringAttr>(
+              gpu::getDefaultGpuBinaryAnnotation());
+          if (!binaryAttr) {
+            gpuMod.emitOpError() << "missing gpu.binary attribute";
+            return;
+          }
 
-        gpuMods.push_back(gpuMod);
-
-        // apply target spec to original func
-        gpuMod.walk([&](LLVM::LLVMFuncOp func) {
-          if (auto attr = func->getAttrOfType<SymbolRefAttr>("original_func")) {
-            if (auto miopenFunc = mod.lookupSymbol<func::FuncOp>(attr)) {
-              std::vector<NamedAttribute> attributes{
-                  b.getNamedAttr("type", b.getStringAttr("gpu")),
-                  b.getNamedAttr("arch", gpuMod->getAttr("arch")),
-                  b.getNamedAttr("grid_size", func->getAttr("grid_size")),
-                  b.getNamedAttr("block_size", func->getAttr("block_size")),
-                  b.getNamedAttr("binary", binaryAttr)};
-
-              miopenFunc->setAttr(
-                  "targets", b.getArrayAttr({b.getDictionaryAttr(attributes)}));
+          gpuMods.push_back(gpuMod);
+
+          // apply target spec to original func
+          gpuMod.walk([&](LLVM::LLVMFuncOp func) {
+            if (auto attr =
+                    func->getAttrOfType<SymbolRefAttr>("original_func")) {
+              if (auto kernelFunc = mod.lookupSymbol<func::FuncOp>(attr)) {
+                std::vector<NamedAttribute> attributes{
+                    b.getNamedAttr("type", b.getStringAttr("gpu")),
+                    b.getNamedAttr("arch", gpuMod->getAttr("arch")),
+                    b.getNamedAttr("grid_size", func->getAttr("grid_size")),
+                    b.getNamedAttr("block_size", func->getAttr("block_size")),
+                    b.getNamedAttr("binary", binaryAttr)};
+
+                kernelImpls[kernelFunc].push_back(
+                    b.getDictionaryAttr(attributes));
```

Comment:
I'm not sure this here bit of nested attribute spaghetti is the right approach.  I'd at least like to see a `GpuKernel` attribute that combines grid_size, block_size, and binary. Then we could map from `(SystemType, StringAttr) -> Attribute` , where for a GPU kernel, the attribute would be a `GPUKernel` attribute.

This would also probably mean making `SystemType` a tablegen'd enum, so you can have a `SystemTypeAttr` but I'd say that's a solid thing anyway.